File: 1277-count-square-submatrices-with-all-ones/1277-count-square-submatrices-with-all-ones.cpp
class Solution {
// TC & SC: O(mn)
public:
    int countSquares(vector<vector<int>>& matrix) {
        int row = matrix.size(), col = matrix[0].size();
        vector<vector<int>> dp(row, vector<int> (col, 0));

        for(int i = 0; i < row; ++i)
            dp[i][0] = matrix[i][0];
        for(int j = 0; j < col; ++j)
            dp[0][j] = matrix[0][j];
        
        for(int i = 1; i < row; ++i) {
            for(int j = 1; j < col; ++j) {
                if(matrix[i][j] == 0)
                    dp[i][j] = 0;
                else 
                    dp[i][j] = 1 + min(dp[i - 1][j], min(dp[i - 1][j - 1], dp[i][j - 1]));
            }
        }
        
        int sum = 0;
        for(int i = 0; i < row; ++i)
            for(int j = 0; j < col; ++j)
                sum += dp[i][j];
        
        return sum;
    }
};

File: 1277-count-square-submatrices-with-all-ones/1277-count-square-submatrices-with-all-ones.py
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        row = len(matrix)
        col = len(matrix[0])
        
        dp = [[0] * col for _ in range(row)]
        
        for i in range (0, row):
            dp[i][0] = matrix[i][0]
        for j in range (0, col):
            dp[0][j] = matrix[0][j]
            
        for i in range (1, row):
            for j in range (1, col):
                if matrix[i][j] == 0:
                    dp[i][j] = 0
                else:
                    dp[i][j] = 1 + min(dp[i - 1][j - 1], min(dp[i - 1][j], dp[i][j - 1]))
        
        sum = 0
        for i in range (0, row):
            for j in range (0, col):
                sum += dp[i][j]
        return sum